### ISSUE TITLE
The results obtained from torch.svd()-based pseudoinversion and torch.linalg.pinv() differ, and they exhibit different gradient behaviors during backpropagation.

### ISSUE THREAD

**Issue by manmanjun** [pytorch/pytorch#110009]
### 🐛 Describe the bug

```python
import torch
A = torch.rand(5,10)
A.requires_grad=True

k=2

####### torch.linalg.pinv
b = A.t()@A
u,d,v = torch.svd(b)
pinv = torch.linalg.pinv(b,rtol = (d[k]/d[0]).detach())
print(pinv.sum())
loss = pinv.sum()
loss.backward()
print(A.grad)
A.grad.zero_()

####### torch.svd
b = A.t()@A
u,d,v = torch.svd(b)
d_new = torch.zeros_like(d)
d_new[:k] = 1 / d[:k]
pinv_svd = u@torch.diag(d_new)@v.t()
print(pinv_svd.sum())
loss = pinv_svd.sum()
loss.backward()
print(A.grad)
```

### Versions

PyTorch version: 1.10.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.17

Python version: 3.7.11 (default, Jul 27 2021, 14:32:16)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.15.0-83-generic-x86_64-with-debian-buster-sid
Is CUDA available: True
CUDA runtime version: 11.3.109
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: Tesla V100S-PCIE-32GB
GPU 1: Tesla V100S-PCIE-32GB
GPU 2: Tesla V100S-PCIE-32GB
GPU 3: Tesla V100S-PCIE-32GB
GPU 4: Tesla V100S-PCIE-32GB
GPU 5: Tesla V100S-PCIE-32GB
GPU 6: Tesla V100S-PCIE-32GB
GPU 7: Tesla V100S-PCIE-32GB

Nvidia driver version: 470.82.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              96
On-line CPU(s) list: 0-95
Thread(s) per core:  2
Core(s) per socket:  24
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
Stepping:            7
CPU MHz:             1171.256
CPU max MHz:         4000.0000
CPU min MHz:         1000.0000
BogoMIPS:            4400.00
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            36608K
NUMA node0 CPU(s):   0-23,48-71
NUMA node1 CPU(s):   24-47,72-95
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke avx512_vnni md_clear flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.21.2
[pip3] torch==1.10.0
[pip3] torchelastic==0.2.0
[pip3] torchtext==0.11.0
[pip3] torchvision==0.11.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               ha36c431_9    nvidia
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.3.0           h06a4308_520  
[conda] mkl-service               2.4.0            py37h7f8727e_0  
[conda] mkl_fft                   1.3.1            py37hd3c417c_0  
[conda] mkl_random                1.2.2            py37h51133e4_0  
[conda] numpy                     1.21.2           py37h20f2e39_0  
[conda] numpy-base                1.21.2           py37h79a1101_0  
[conda] pytorch                   1.10.0          py3.7_cuda11.3_cudnn8.2.0_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchelastic              0.2.0                    pypi_0    pypi
[conda] torchtext                 0.11.0                     py37    pytorch
[conda] torchvision               0.11.0               py37_cu113    pytorch

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @xwang233 @Lezcano

**Comment by tringwald** [comment]
The [docs of `torch.svd`](https://pytorch.org/docs/stable/generated/torch.svd.html) mention that 
> If the distance between any two singular values is close to zero, the gradients with respect to U and V will be numerically unstable

>The same happens when the matrix has small singular values

This seems to be the case in your provided code. I've written a small helper script based on your code to obtain those values:

```python
import torch

print(torch.__version__)
torch.manual_seed(1)
A = torch.rand(5, 10)
A.requires_grad = True
b = A.t() @ A
u, d, v = torch.svd(b)
sv_diffs = ((d.view(-1, 1) - d.view(1, -1)) +
            (torch.eye(d.shape[0]) * 99999))
print("Min singular value diff", sv_diffs.abs().min().item())
print("Smallest singular value", d.min().item())
k = 2

# torch.linalg.pinv
pinv = torch.linalg.pinv(b, rtol=(d[k] / d[0]).detach())
loss = pinv.sum()
loss.backward(retain_graph=True)
pinv_grad = A.grad.clone().detach()
A.grad.zero_()

# torch.svd
d_new = torch.zeros_like(d)
d_new[:k] = 1 / d[:k]
pinv_svd = u @ torch.diag(d_new) @ v.t()
loss = pinv_svd.sum()
loss.backward(retain_graph=True)
svd_grad = A.grad.clone().detach()
A.grad.zero_()

print("Grad max diff", (pinv_grad - svd_grad).abs().max().item())
print("Grad median diff", (pinv_grad - svd_grad).abs().median().item())
print("Max diff", (pinv.sum() - pinv_svd.sum()).abs().max().item())
```
which results in
```
2.0.1
Min singular value diff 3.357543576498756e-08
Smallest singular value 4.1014093454805334e-08
Grad max diff 0.011817499995231628
Grad median diff 0.0035834982991218567
Max diff 1.1920928955078125e-07
```

i.e. very small singular values and a somewhat larger grad difference. The difference of the actual result seems to be in a reasonable range of `1e-07`.

Can you run this code and post the output?

**Comment by manmanjun** [comment]
> The [docs of `torch.svd`](https://pytorch.org/docs/stable/generated/torch.svd.html) mention that
> 
> > If the distance between any two singular values is close to zero, the gradients with respect to U and V will be numerically unstable
> 
> > The same happens when the matrix has small singular values
> 
> This seems to be the case in your provided code. I've written a small helper script based on your code to obtain those values:
> 
> ```python
> import torch
> 
> print(torch.__version__)
> torch.manual_seed(1)
> A = torch.rand(5, 10)
> A.requires_grad = True
> b = A.t() @ A
> u, d, v = torch.svd(b)
> sv_diffs = ((d.view(-1, 1) - d.view(1, -1)) +
>             (torch.eye(d.shape[0]) * 99999))
> print("Min singular value diff", sv_diffs.abs().min().item())
> print("Smallest singular value", d.min().item())
> k = 2
> 
> # torch.linalg.pinv
> pinv = torch.linalg.pinv(b, rtol=(d[k] / d[0]).detach())
> loss = pinv.sum()
> loss.backward(retain_graph=True)
> pinv_grad = A.grad.clone().detach()
> A.grad.zero_()
> 
> # torch.svd
> d_new = torch.zeros_like(d)
> d_new[:k] = 1 / d[:k]
> pinv_svd = u @ torch.diag(d_new) @ v.t()
> loss = pinv_svd.sum()
> loss.backward(retain_graph=True)
> svd_grad = A.grad.clone().detach()
> A.grad.zero_()
> 
> print("Grad max diff", (pinv_grad - svd_grad).abs().max().item())
> print("Grad median diff", (pinv_grad - svd_grad).abs().median().item())
> print("Max diff", (pinv.sum() - pinv_svd.sum()).abs().max().item())
> ```
> 
> which results in
> 
> ```
> 2.0.1
> Min singular value diff 3.357543576498756e-08
> Smallest singular value 4.1014093454805334e-08
> Grad max diff 0.011817499995231628
> Grad median diff 0.0035834982991218567
> Max diff 1.1920928955078125e-07
> ```
> 
> i.e. very small singular values and a somewhat larger grad difference. The difference of the actual result seems to be in a reasonable range of `1e-07`.
> 
> Can you run this code and post the output?

```
1.12.1
Min singular value diff 3.4353703881606634e-08
Smallest singular value 7.28376470249259e-09
Grad max diff 0.011817552149295807
Grad median diff 0.0035834982991218567
Max diff 5.960464477539063e-08
```
Thank you for your response. I'm just curious why there is a difference between torch.svd() and torch.linalg.pinv(), especially since it seems that torch.linalg.pinv() simply calls torch.svd() in the csrc [source code](https://github.com/pytorch/pytorch/blob/664058fa83f1d8eede5d66418abff6e20bd76ca8/aten/src/ATen/native/LinearAlgebra.cpp#L372).

**Comment by tringwald** [comment]
The `torch.linalg.pinv` code seems to calculate [`V @ diag(S) @ U.T`](https://github.com/pytorch/pytorch/blob/664058fa83f1d8eede5d66418abff6e20bd76ca8/aten/src/ATen/native/LinearAlgebra.cpp#L376), while your code above calculates `u @ torch.diag(d_new) @ v.t()`. When I change it to the torch version, the output difference goes to 0.

**Comment by manmanjun** [comment]
> The `torch.linalg.pinv` code seems to calculate [`V @ diag(S) @ U.T`](https://github.com/pytorch/pytorch/blob/664058fa83f1d8eede5d66418abff6e20bd76ca8/aten/src/ATen/native/LinearAlgebra.cpp#L376), while your code above calculates `u @ torch.diag(d_new) @ v.t()`. When I change it to the torch version, the output difference goes to 0.

the `torch.diag(d_new)` means the operation of `rtol` in  `torch.linalg.pinv`. You can also run this code：
```python
import torch

A = torch.rand(5,10).double()
B = A.detach().clone()
B[0][0] = B[0][0] + 0.00001
A.requires_grad=True
B.requires_grad=True
k=2

####### torch.linalg.pinv
b = A.t()@A
u,d,v = torch.svd(b)
pinv = torch.linalg.pinv(b,rtol = (d[k]/d[0]).detach())
loss = pinv.sum()
loss.backward()
grad_pinv = A.grad[0][0].clone().detach()
A.grad.zero_()

####### torch.svd
b = A.t()@A
u,d,v = torch.svd(b)
d_new = torch.zeros_like(d)
d_new[:k] = 1 / d[:k]
pinv_svd = u@torch.diag(d_new)@v.t()
loss = pinv_svd.sum()
loss.backward()
grad_svd = A.grad[0][0].clone().detach()
print('grad_pinv', grad_pinv)
print('grad_svd', grad_svd)
print("Grad diff", (grad_pinv - grad_svd).abs().item())
```
```
grad_pinv tensor(-0.0371, dtype=torch.float64)
grad_svd tensor(-0.0314, dtype=torch.float64)
Grad diff 0.005700894257729683
```

**Comment by lezcano** [comment]
The svd is not differentiable at singular matrices, so you cannot use backward on it for those values (the chain rule does not apply in this case).

 `pinv` is not differentiable either at these values (it's not even continuous!!), but you can define a derivative of sorts defining its derivative as the derivative on the strata defined by the matrices of constant rank. A formula to compute this can be found at
>The Differentiation of Pseudo-Inverses and Nonlinear Least Squares Problems Whose Variables Separate. Author(s): G. H. Golub and V. Pereyra. Source: SIAM Journal on Numerical Analysis, Vol. 10, No. 2 (Apr., 1973), pp. 413-432

and that's the formula we implement in PyTorch.

**Comment by nikitaved** [comment]
These inputs are fine, the problem is that the forward passes are not equivalent. Have a look at. how `atol/rtol` are being processed:
https://github.com/pytorch/pytorch/blob/132a138a01806b45bb050cbcacbaa782fcf2e2ae/aten/src/ATen/native/LinearAlgebra.cpp#L515-L524.

You can see that a max singular value is used for filtering, so `pinv_B` and `pinv_svd_B` are not "equivalent".

**Comment by manmanjun** [comment]
> rtol

Thank you for your response. I understand that `a max singular value is used for filtering`.  `rtol` in `torch.linalg.pinv` is actually used to filter out smaller singular values. For convenience, I pre-designed `k=2` in the code to ensure that both functions select the same singular values when obtaining the pseudo-inverse.

**Comment by lezcano** [comment]
For reference, no, I don't think there's any bug here, just a small misconception :)
>My only question for now is why the two functions yield different gradients if they use the same pre-defined formula.

Calling backwards will just yield reliable gradients when the functions are differentiable (as it uses the chain rule, and this is only valid when the functions are differentiable). This is documented in
https://pytorch.org/docs/stable/notes/autograd.html#gradients-for-non-differentiable-functions
and
https://pytorch.org/docs/stable/notes/numerical_accuracy.html#linear-algebra-torch-linalg

Closing this, as this is expected and documented behaviour. Feel free to open a thread in [the formus](https://discuss.pytorch.org/)
if you still have any questions :)
